pgjdbc/src/main/java/com/yugabyte/ysql/LoadBalanceManager.java

```diff
@@ -0,0 +1,459 @@
+    Properties props = loadBalanceProperties.getOriginalProperties();
+    String url = loadBalanceProperties.getStrippedURL();
+
+    if (!checkAndRefresh(loadBalanceProperties, lb, user, dbName)) {
+      return null;
+    }
+
+    List<String> failedHosts = new ArrayList<>();
+    String chosenHost = lb.getLeastLoadedServer(true, failedHosts);
+    PgConnection newConnection;
+    SQLException firstException = null;
+    while (chosenHost != null) {
+      try {
+        props.setProperty("PGHOST", chosenHost);
+        props.setProperty("PGPORT", String.valueOf(LoadBalanceManager.getPort(chosenHost)));
+        newConnection = new PgConnection(hostSpecs(props), user, dbName, props, url);
+        newConnection.setLoadBalancer(lb);
+        LOGGER.fine("Created connection to " + chosenHost);
+        return newConnection;
+      } catch (SQLException ex) {
+        // Let the refresh be forced the next time it is tried. todo Is this needed?
+        forceRefreshOnce = true;
+        failedHosts.add(chosenHost);
+        LoadBalanceManager.decrementConnectionCount(chosenHost);
+        if (PSQLState.CONNECTION_UNABLE_TO_CONNECT.getState().equals(ex.getSQLState())) {
+          if (firstException == null) {
+            firstException = ex;
+          }
+          LOGGER.fine("couldn't connect to " + chosenHost + ", adding it to failed host list");
+          LoadBalanceManager.markAsFailed(chosenHost);
+        } else {
+          // Log the exception. Consider other failures as temporary and not as serious as
+          // PSQLState.CONNECTION_UNABLE_TO_CONNECT. So for other failures it will be ignored
+          // only in this attempt, instead of adding it in the failed host list of the
+          // load balancer itself because it won't be tried till the next refresh happens.
+          LOGGER.warning("got exception " + ex.getMessage() + " while connecting to " + chosenHost);
+        }
+      } catch (Throwable e) {
+        LOGGER.fine("Received Throwable: " + e);
+        LoadBalanceManager.decrementConnectionCount(chosenHost);
+        throw e;
+      }
+      chosenHost = lb.getLeastLoadedServer(false, failedHosts);
+    }
+    return null;
+  }
+
+  private static synchronized boolean checkAndRefresh(LoadBalanceProperties loadBalanceProperties,
+      LoadBalancer lb, String user, String dbName) {
+    if (needsRefresh(lb.getRefreshListSeconds())) {
+      Properties props = loadBalanceProperties.getOriginalProperties();
+      String url = loadBalanceProperties.getStrippedURL();
+      HostSpec[] hspec = hostSpecs(props);
+
+      ArrayList<String> hosts = getAllAvailableHosts(new ArrayList<>());
+      while (true) {
+        try {
+          if (controlConnection == null || controlConnection.isClosed()) {
+            controlConnection = new PgConnection(hspec, user, dbName, props, url);
+          }
+          refresh(controlConnection, lb.getRefreshListSeconds());
+          controlConnection.close();
```

Comment:
Here the control connection is getting closed so will we create a new control connection each time refresh happens?

Reply:
Good catch, thanks. No, control connection will not be created for every refresh. This was meant to be removed in my last commit `f10f7f1008d2d2aa88b8c780e88ee271ac8e4430`. Will remove it.